### PR TITLE
gh-125196: Use PyUnicodeWriter in HAMT

### PR DESCRIPTION
**Pull request by vstinner**
Replace the private _PyUnicodeWriter API with the public PyUnicodeWriter API in hamt.c.

Remove _hamt_dump_format(): replace it with PyUnicodeWriter_Format().

Use PyUnicodeWriter_WriteUTF8() to write constant strings.

* Issue: gh-125196

### CODE REVIEW

File: Python/hamt.c [hunk 1]
```diff
@@ -1558,8 +1536,8 @@ hamt_node_collision_dump(PyHamtNode_Collision *node,
         goto error;
     }
 
-    if (_hamt_dump_format(writer, "CollisionNode(size=%zd id=%p):\n",
-                          Py_SIZE(node), node))
+    if (PyUnicodeWriter_Format(writer, "CollisionNode(size=%zd id=%p):\n",
+                          Py_SIZE(node), node) < 0)
```
Comment (picnixz) [inline Python/hamt.c]:
```suggestion
    if (PyUnicodeWriter_Format(writer, "CollisionNode(size=%zd id=%p):\n",
                          	   Py_SIZE(node), node) < 0)
```

File: Python/hamt.c [hunk 2]
```diff
@@ -1181,7 +1158,8 @@ hamt_node_bitmap_dump(PyHamtNode_Bitmap *node,
     if (tmp2 == NULL) {
         goto error;
     }
-    if (_hamt_dump_format(writer, "bitmap=%S id=%p):\n", tmp2, node)) {
+    if (PyUnicodeWriter_Format(writer, "bitmap=%S id=%p):\n",
+                               tmp2, node) < 0) {
```
Comment (picnixz) [inline Python/hamt.c]:
```suggestion
    if (PyUnicodeWriter_Format(writer, "bitmap=%S id=%p):\n",
                               tmp2, node) < 0)
    {
```

**Review by 1st1 — APPROVED** [review]

**Comment by nineteendo** [comment]
You need to update the branch first.

**Comment by vstinner** [comment]
> You need to update the branch first.

Right, done.

**Pull request merged**